### PR TITLE
Fix new api implementation

### PR DESCRIPTION
**Pull request by yurybotov**
Replace Octoliner implementation to GpioExpander based with last API.md agreement. Corrected after 'live' test.

### CODE REVIEW

File: src/Octoliner.h [hunk 1]
```diff
@@ -1,102 +1,48 @@
...
-    void writeCmd16BitData(IOcommand command, uint16_t data);
-    void writeCmd8BitData(IOcommand command, uint8_t data);
-    void writeCmd(IOcommand command, bool sendStop = true);
-    int read16Bit();
-    uint32_t read32bit();
+    const uint8_t _IRLedsPin = 9;
+    const uint8_t _sensePin = 0;
+    const uint8_t _sensorPinMap[8] = {
+        4, 5, 6, 8, 7, 3, 2, 1
+    };
```
Comment (nkrkv) [inline src/Octoliner.h]: No need to keep these in RAM for every instance of the class. Better move them to C++ and make `constexpr`

File: src/Octoliner.h [hunk 2]
```diff
@@ -1,102 +1,48 @@
...
-#ifndef INPUT_PULLDOWN
-#define INPUT_PULLDOWN 0x3
-#endif
-//INPUT_PULLDOWN
-
-#ifndef DEFAULT_GPIOEXP_ADDR
-#define DEFAULT_GPIOEXP_ADDR 42
-#endif
-// DEFAULT_GPIOEXP_ADDR
+class Octoliner : public GpioExpander {
```
Comment (nkrkv) [inline src/Octoliner.h]:
```suggestion
class Octoliner : private GpioExpander {
```

We don’t want to expose the internal details of `GpioExpander`. Don’t forget to expose `setAddress` and `saveAddress` though.

File: src/Octoliner.cpp [hunk 3]
```diff
@@ -10,292 +10,114 @@
...
 
-void Octoliner::writeCmdPin(IOcommand command, uint8_t pin, bool sendStop) {
-    Wire.beginTransmission(_i2caddress);
-    Wire.write((uint8_t)command);
-    Wire.write(pin);
-    Wire.endTransmission(sendStop);
+void Octoliner::begin(TwoWire* wire) {
+    wire->begin();
+    GpioExpander::begin(wire);
+    GpioExpander::pwmFreq(8000); // ~ 250 different pwm values
```
Comment (nkrkv) [inline src/Octoliner.cpp]:
```suggestion
    GpioExpander::pwmFreq(8000); // ~ 250 pwm levels
```

…and why we ever need to touch the PWM settings?

**Review by nkrkv — CHANGES_REQUESTED** [review]
Nice! A few points below

File: src/Octoliner.h (hunk 1 above)
Comment (yurybotov) [reply src/Octoliner.h]: For single const - ok, but for array - it is bad idea (linker cant find const expression). I made "static const _sensorPinMap[8];"

File: src/Octoliner.h (hunk 2 above)
Comment (yurybotov) [reply src/Octoliner.h]: ...and add public definitions for change and saveAddress. Done.

File: src/Octoliner.cpp (hunk 3 above)
Comment (yurybotov) [reply src/Octoliner.cpp]: Default frequency in stm32f030 firmware is 1kHz. It is good for servo driving, but for sensitivity level set it is slow - will be more noise.

**Review by nkrkv — APPROVED** [review]
:+1:

File: src/Octoliner.h [hunk 4]
```diff
@@ -1,102 +1,49 @@
...
-    int analogRead(int pin);
-    uint32_t getUID();
-    int digitalReadPort();
-    void digitalWritePort(uint16_t value);
-    void pinModePort(uint16_t value, uint8_t mode);
-    void adcSpeed(uint8_t speed);
-    void setSensitivity(uint8_t sense);
-    void setBrightness(uint8_t brightness);
-    float mapLine(int binaryLine[8]);
+    void saveAddr(void);
```
Comment (nkrkv) [inline src/Octoliner.h:37]: Ouch… These are `changeAddress` and `saveAddress` in `API.md`. Suggest renaming to match the docs.

**Pull request merged**